GVFS/GVFS.UnitTests.Windows/Windows/Upgrader/UpgradeOrchestratorTests.cs

```diff
@@ -0,0 +1,243 @@
+                downloadNewestVersion: false,
+                installNewestVersion: false,
+                cleanup: true);
+
+            this.orchestrator.ExitCode.ShouldEqual(ReturnCode.GenericError);
+        }
+
+        [TestCase]
+        public void DownloadFails()
+        {
+            this.MoqUpgrader.Setup(upgrader => upgrader.TryDownloadNewestVersion(out It.Ref<string>.IsAny))
+                .Callback(new TryDownloadNewestVersionCallback((out string delegateMessage) =>
+                {
+                    delegateMessage = "Download error.";
+                }))
+                .Returns(false);
+
+            this.orchestrator.Execute();
+
+            this.VerifyOrchestratorInvokes(
+                upgradeAllowed: true,
+                queryNewestVersion: true,
+                downloadNewestVersion: true,
+                installNewestVersion: false,
+                cleanup: true);
+
+            this.VerifyOutput("ERROR: Download error.");
+
+            this.orchestrator.ExitCode.ShouldEqual(ReturnCode.GenericError);
+        }
+
+        [TestCase]
+        public void RunInstallerFails()
+        {
+            this.MoqUpgrader.Setup(upgrader => upgrader.TryRunInstaller(It.IsAny<InstallActionWrapper>(), out It.Ref<string>.IsAny))
+                .Callback(new TryRunInstallerCallback((InstallActionWrapper installActionWrapper, out string delegateMessage) =>
+                {
+                    delegateMessage = "Installer error.";
+                }))
+                .Returns(false);
+
+            this.orchestrator.Execute();
+
+            this.VerifyOrchestratorInvokes(
+                upgradeAllowed: true,
+                queryNewestVersion: true,
+                downloadNewestVersion: true,
+                installNewestVersion: true,
+                cleanup: true);
+
+            this.VerifyOutput("ERROR: Installer error.");
+
+            this.orchestrator.ExitCode.ShouldEqual(ReturnCode.GenericError);
+        }
+
+        public Mock<IProductUpgrader> DefaultUpgrader()
+        {
+            Mock<IProductUpgrader> mockUpgrader = new Mock<IProductUpgrader>();
+
+            mockUpgrader.Setup(upgrader => upgrader.UpgradeAllowed(out It.Ref<string>.IsAny))
+                .Callback(new UpgradeAllowedCallback((out string delegateMessage) =>
+                {
+                    delegateMessage = string.Empty;
+                }))
+                .Returns(true);
+
+            string message = string.Empty;
+            mockUpgrader.Setup(upgrader => upgrader.TryDownloadNewestVersion(out It.Ref<string>.IsAny)).Returns(true);
+            mockUpgrader.Setup(upgrader => upgrader.TryRunInstaller(It.IsAny<InstallActionWrapper>(), out message)).Returns(true);
+            mockUpgrader.Setup(upgrader => upgrader.TryCleanup(out It.Ref<string>.IsAny)).Returns(true);
+
+            return mockUpgrader;
+        }
+
+        public void SetUpgradeAvailable(Version newVersion, string error)

```

Comment:
What should `TryQueryNewestVersion` return if there is no upgrade available?  

I ask because it looks like this method is setting whether `TryQueryNewestVersion` succeeds or fails (rather than if it behaves like a new version is available).

Reply:
When no upgrade is available - `TryQueryNewestVersion` returns `true`, `out newVersion` will be set to `null`, `out message` will be set to "`... you're all caught up on upgrades..`".

Reply:
I do think the meaning of the return value for `TryQueryNewestVersion` method can be ambiguous or even initially unexpected, but I do think it follows the semantics for our `Try....` methods (I think...). We successfully queried the latest version - it just happens there is no newer version.

I had to read through this a couple of times to make sure I understood how to interpret the return value. Outside of this PR, maybe we could make this less confusing, as I do expect this to cause confusion in the future...